File: Server/Src/Drones/DroneHandler.cs
using System.Net.WebSockets;
using System.Text.Json;

public class DroneHandler
{
    private static DroneHandler instance;
    private readonly DroneManager droneManager = DroneManager.GetInstance();

    private DroneHandler() { }

    public static DroneHandler GetInstance()
    {
        if (instance == null)
            instance = new DroneHandler();
        return instance;
    }
    

    public void HandleRequestDronesInitData(WebSocket connection, JsonElement data)
    {
        try
        {
            // צור UUID חדש
            Guid uuid = Guid.NewGuid();
            string uuidString = uuid.ToString();

            // צור רחפן חדש עם ID ודיפולט
            var defaultPosition = new GeoPoint(34.78217676812864, 32.02684069644974, 160);
            var defaultTrajectory = new TrajectoryPoint(defaultPosition, 0, 0);
            var drone = new Drone(uuidString, defaultTrajectory);

            // הוסף ל־DroneManager
            droneManager.TryAddDrone(drone);

            // שלח חזרה ללקוח DRONEINITDATA רק עם ה-ID
            var initData = new
            {
                yourDroneId = uuidString
            };
            string json = WebSocketServer.prepareMessageToClient(S2CMessageType.DroneInitData, initData);
            WebSocketServer.SendMsgToClient(connection, json);
        }
        catch (Exception ex)
        {
            Console.WriteLine("Error in HandleRequestDronesInitData: " + ex.Message);
        }
    }


    public void HandleAddDrone(JsonElement data)
    {
        try
        {
            Drone drone = data.Deserialize<Drone>();
            if (drone == null)
                throw new Exception("Deserialization returned null");

            Guid uuid = Guid.NewGuid();
            string uuidString = uuid.ToString();
            drone.id = uuidString;

            bool added = droneManager.TryAddDrone(drone);
            if (added)
            {
                Console.WriteLine($"{drone.id} - Drone added successfully.");
            }
            else
            {
                Console.WriteLine($"{drone.id} - Failed to add drone.");
                SendDroneError($"{drone.id} - Failed to add drone.");
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine("Error in HandleAddDrone: " + ex.Message);
        }
    }

    public void HandleRemoveDrone(JsonElement data)
    {
        try
        {
            Drone drone = data.Deserialize<Drone>();
            if (drone == null)
                throw new Exception("Deserialization returned null");

            bool removed = droneManager.TryRemoveDrone(drone.id);
            if (removed)
            {
                Console.WriteLine($"{drone.id} - Drone removed successfully.");
                SendRemoveDrone(drone);
            }
            else
            {
                Console.WriteLine($"{drone.id} - Failed to remove drone.");
                SendDroneError($"{drone.id} - Failed to remove drone.");
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine("Error in HandleRemoveDrone: " + ex.Message);
        }
    }

    public void HandleUpdateDrone(JsonElement data)
    {
        try
        {
            Drone drone = data.Deserialize<Drone>();
            if (drone == null)
                throw new Exception("Deserialization returned null");

            bool updated = droneManager.TryUpdateDrone(drone.id, drone);
            if (updated)
            {
                Console.WriteLine($"{drone.id} - Drone updated successfully.");
                SendUpdateDrone(drone);
            }
            else
            {
                Console.WriteLine($"{drone.id} - Failed to update drone.");
                SendDroneError($"{drone.id} - Failed to update drone.");
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine("Error in HandleUpdateDrone: " + ex.Message);
        }
    }

    public void SendRemoveDrone(Drone drone)
    {
        string data = WebSocketServer.prepareMessageToClient(S2CMessageType.RemoveDrone, drone);
        WebSocketServer.SendMsgToClients(data);
    }

    public void SendUpdateDrone(Drone drone)
    {
        string data = WebSocketServer.prepareMessageToClient(S2CMessageType.UpdateDrone, drone);
        WebSocketServer.SendMsgToClients(data);
    }

    public void SendDroneError(string errorMsg)
    {
        var err = new { errorMsg };
        string data = WebSocketServer.prepareMessageToClient(S2CMessageType.DroneError, err);
        WebSocketServer.SendMsgToClients(data);
    }
}

File: ReactClient/src/FreeFlightMode/Drones/DroneHandler.ts
import type { Drone } from "../../Messages/AllTypes";
import { DroneEntityManager } from "./DroneEntityManager";
import * as Cesium from "cesium";
import { toast } from "react-toastify";

export class DroneHandler {
  private static instance: DroneHandler | null = null;
  private droneEntityManager: DroneEntityManager;

  private constructor(viewer: Cesium.Viewer) {
    this.droneEntityManager = DroneEntityManager.getInstance(viewer);
  }

  public static getInstance(viewer?: Cesium.Viewer): DroneHandler {
    if (!this.instance) {
      if (!viewer) {
        throw new Error("Viewer must be provided for the first getInstance call");
      }
      this.instance = new DroneHandler(viewer);
    }
    return this.instance;
  }

  public HandleAddDrone(data: any) {
    try {
      const drone = data as Drone;
      const isAdded = this.droneEntityManager.tryAddDrone(drone);
      if (isAdded) console.log(`Drone ${drone.id} added successfully.`);
      else console.log(`Failed to add drone ${drone.id}.`);
    } catch (err) {
      console.error("Data could not be parsed to Drone:", err);
    }
  }

  public HandleRemoveDrone(data: any) {
    try {
      const drone = data as Drone;
      const isRemoved = this.droneEntityManager.removeDrone(drone.id);
      if (isRemoved) console.log(`Drone ${drone.id} removed successfully.`);
      else console.log(`Failed to remove drone ${drone.id}.`);
    } catch (err) {
      console.error("Data could not be parsed to Drone:", err);
    }
  }

  public HandleUpdateDrone(data: any) {
    try {
      const drone = data as Drone;
      const isEdited = this.droneEntityManager.editDrone(drone);
      if (isEdited) console.log(`Drone ${drone.id} updated successfully.`);
      else console.log(`Failed to update drone ${drone.id}.`);
    } catch (err) {
      console.error("Data could not be parsed to Drone:", err);
    }
  }

  public HandleDroneError(data: any) {
    try {
      const errorMsg = (data as { errorMsg: string }).errorMsg;
      console.log("Drone error from server: " + errorMsg);
      toast.error(errorMsg);
    } catch {
      console.error("Data could not be parsed to DroneError");
    }
  }

  public getDroneEntity(droneId: string): Cesium.Entity | null {
    return this.droneEntityManager.getDroneEntity(droneId);
  }

  public getAllDroneIds(): string[] {
    return this.droneEntityManager.getAllDroneIds();
  }

  public clearAllDrones() {
    this.droneEntityManager.clearAllDrones();
  }
}
